### ISSUE TITLE
Destroy Action

### ISSUE THREAD

**Issue by Nalum** [hashicorp/terraform#649]
Is there a way to have a resource run a command before it is destroyed?

Example: I have an `aws_instance` that when provisioned was added to a chef servers node and client list, when this resource is destroyed I want to remove it from the node and client list.

**Comment by bitglue** [comment]
I would think a much cleaner way to address this problem would be creating a terraform resource that represents the entry in the Chef client list. Provisioners are a hack at best, and this highlights the reason why. If you think of your infrastructure as a state machine, then Terraform is based on modelling the desired state and Terraform then selects the right CRUD operations from the providers to make the edges to get to that state. Provisioners don't model desired state, they model edges. Consequently Terraform doesn't know anything about how that provisioner altered state, and you have a bad time.

If you want a tool that models edges and not state, that tool already exists. It's called Ansible. Or Chef. Or Puppet. There's no reason you can't do both. I don't see any reason to dilute Terraform's model to support a use case that's already well-addressed by other tools.

**Comment by knuckolls** [comment]
I agree with you about the possibility of using custom terraform resources for things like the "clean up the chef client on instance destruction". Although, I do respectfully disagree with you about the need for a before-destroy provisioner specifically. I have a blog post largely written up about this idea as a part of a grander scheme of other things we've been working on but here's the part that's pertinent to this issue specifically.

Assume you have a cluster of stateful machines. An elasticsearch cluster for example. Anything that is clustered by default with a >1 replication factor fits this model quite nicely actually. Now, assume that we want to cut a new server image with packer and roll the cluster so that all the machines are on the new image. There are a variety of good reasons to do this. Right now Terraform does not have a safe way to drain old AMIs one by one. It can either delete the whole cluster and re-create it, or stand up the new nodes and then delete the old nodes. I believe we require stronger guarantees about how our stateful instances get destroyed. For example, we should never delete all the nodes that contain a given block worth of data until we have confirmed that the data has been replicated to the new set of instances. Create before destroy doesn't give us that assurance. It merely waits for the new instance to be up before terminating the old instance. 

Terraform is very close to providing the correct primitives for such graceful draining of stateful instances. It requires two things: the parallelism semaphore deep within terraform/context.go should be exposed up to the command line UI as a "--parallelism" flag so you can enforce that a specific apply is done one node at a time. Right now parallelism is just hardcoded to a default of 10 items at a time if it isn't overridden and the current implementation of what sets up the ContextOpts doesn't override it as far as I tell. The second thing is that stateful instances controlled by terraform require the need to gracefully decomission themselves and then spin until their blocks have been replicated to the other nodes before starting the next destroy. This decomission and blocking dance can be done within before-destroy provisioners on the instance resource. Either me or someone on our team will be PRing some of this work as we get to it and I'd like to hopefully come to some sort of agreement on how it could be done that the community and Hashicorp is comfortable with.

Furthermore, tainting a stateful node may not be strong enough of an assurance when it comes to the automated destruction of stateful instances. I believe a new "decomission-on-delete" flag may need to be created so that we can provide the worst case assurance that you can flag specific instances as ones that we want humans to follow up and destroy, just in case. Decomissioning a resource would merely take it out of the statefile, or move it into a decomissioned section and whine about how the user needs to clean these up on every terraform plan. This is because we don't want Terraform to accidentally cause a "we need to recover from cold backups" scenario because too many nodes were deleted too quickly. Decomission-on-delete would put the onus on the operator to be quite certain that the state was replicated off that node before initializing it's destruction. 

I'll eventually publish my thoughts much more extensively about this and related things in a more general sort of way. It did seem prudent to defend my stance that adding an on-destroy provisioner is not a dilution of Terraform's model and is instead an integral piece to the graceful rolling of stateful instances. As always, I'd be willing to make time to discuss these sorts of things on a hangout or move this discussion to it's own dedicated issue if needed. 

Upon re-reading this I realize this diatribe likely falls under the umbrella of "but but someone on the internet is wrong". My apologies if it comes across as gruff in any way. This sort of thing has been a primary concern of mine for awhile now. While this may not be the final solution for graceful rolling of stateful nodes with Terraform, I believe it's one possible way that we just haven't gotten to taking care of yet. :)

**Comment by bitglue** [comment]
> The second thing is that stateful instances controlled by terraform require the need to gracefully decomission themselves and then spin until their blocks have been replicated to the other nodes before starting the next destroy. This decomission and blocking dance can be done within before-destroy provisioners on the instance resource. 

But it can *also* be done by making a resource that represents whatever it is you did when you provisioned the thing. This "provisioning" resource, which might represent some bit of configuration in a cluster (like a Chef client list, or whatever) gets created after the EC2 instance is created, and on destroy, the provisioning resource is destroyed before the instance is destroyed.

> Furthermore, tainting a stateful node may not be strong enough of an assurance when it comes to the automated destruction of stateful instances.

You are absolutely right. "Tainting" is just a half-baked, ambiguously defined state. Since it's also an unknown state, it's impossible to do anything automatically from this state without first resetting to a known state (for example by destroying the resource) or without making restrictions on how provisioners can work (for example requiring that they are idempotent).

> I believe a new "decomission-on-delete" flag may need to be created

But you already *have* that, it's the `Delete` function of a resource provider. If your provisioning is resources, then you don't need the half-baked "tainted" state: you have the real deal, *actual* state, defined exactly as you need it by your provider implementation.

What you are describing with a "deprovisioning" mechanism is CRUD without the R or U, and an incomplete description of the known state (tainting). It's just a half-baked implementation of what Terraform already does, but which can't recover from failures of any of the operations unless you make all operations idempotent, and if you do that, what you've done is implemented Ansible.

**Comment by dalehamel** [comment]
> Not currently, but we want to extend the provisioners to support running them at different points in the lifecycle. e.g. post-create, pre-destroy, etc.

I agree this is a decent approach.

This is definitely related to https://github.com/hashicorp/terraform/issues/386

One thing that @thegedge and I discussed is that provisioners almost seem like they should just be their own resource, rather than a nested attribute of some types of resources.

**Comment by thegedge** [comment]
To explain why I see them as a resource: they add data to the chef server (environment, run list, etc). To me this doesn't seem terribly different from, for example, an `aws_ecs_task_definition` or `consul_keys` resource.

It really is a key component of our infrastructure at the moment, which fits with the definition of a resource. The actual bootstrapping run could be part of the provisioning block, but the client/node setup in the chef server is CRUD-like. Would it make sense to have a chef-server provider?

**Comment by samdunne** [comment]
Has there been any further movement on this? It's becoming a serious issue when coupled with #2720

**Comment by davidblewett** [comment]
+1 on this feature (mainly for ensuring chef stays current). In the mean time, does someone have some kind of script to clean out expired data from chef?

**Comment by icebourg** [comment]
I've taken to using a local provisioner to delete the Chef node and client. This is a horrible hack but it at least lets me create new instances without having to manually delete the old ones. Unfortunately, it only runs when an instance is provisioned, so if you de-provision an instance, you'll still have stale instances in Chef's database. But at least I can recreate instances without worrying what's in Chef.

```
  # Remove the Chef node since a terraform destroy won't. Return true no matter what.
  provisioner "local-exec" {
    command = "knife node delete -y ${var.instance_name}; knife client delete -y ${var.instance_name}; true"
  }
```

**Comment by cleung2010** [comment]
:+1:  for this feature/enhancement

**Comment by Tails** [comment]
+1.

This would come in handy so Terraform could have Digitalocean make a snapshot of a server image instead of deleting it and building from scratch.

**Comment by eidge** [comment]
+1.

Our use case is unregistering app services from consul so that they are taken out of the load balancer before killing the server.

**Comment by davehodgson** [comment]
+1 My use case is that some of the servers are joined to Active Directory domain, it would be very handy to have them delete themselves from the domain pre-destroy. The lack of it means that if I change the server in a way that results in a destroy and recreate the bootstrap to join the domain fails because the account name already exists

**Comment by dhoer** [comment]
+1

**Comment by dhoer** [comment]
https://github.com/hashicorp/terraform/issues/3605

**Comment by avdhoot** [comment]
+1 desperately need this 😀

**Comment by jseaidou** [comment]
has there been any activity on this ?

**Comment by tjlee0909** [comment]
+1 My use case is to pause and wait for the VM to complete and in-progress builds before safely destroying.

**Comment by tjlee0909** [comment]
Is this issue still blocked on a philosophical debate regarding whether this feature is appropriate?

**Comment by dmourati** [comment]
See, e.g.:

http://docs.aws.amazon.com/AutoScaling/latest/DeveloperGuide/AutoScalingGroupLifecycle.html

For how this works in AWS.

**Comment by digitalroadies** [comment]
+1

**Comment by tjlee0909** [comment]
Hey guys, we (Box) is about to start design work on implementing a wrapper around "terraform destroy" for internal use. Are there any updates on this feature request?

**Comment by apparentlymart** [comment]
Don't let my complex proposal discourage a simple solution here... I was trying to collect together a number of use-cases, but my main target was the cases where we use null_resource to trigger actions on updates, rather than destroy actions.

There are a few different ways to attack destroy actions in particular so it would be nice to have a few different approaches to think about.

There are some great ideas in this thread that take a different approach to what I suggested. Trying out some of these ideas with real examples is probably the best way to converge on a solution.

**Comment by eyalzek** [comment]
+1

**Comment by dpnl87** [comment]
I have mainly 2 use cases where this would be really helpful.

1. Windows Server 2012r2 instances that are a domain member. Unjoining them before destroy with a powershell command is quite easy, but I need a trigger somehow.

2. As #3605 mentions a Chef client needs to unregister itself

**Comment by dmourati** [comment]
How about /sbin/halt.local or equivalent? 

https://www.redhat.com/archives/rhl-list/2007-August/msg00314.html

**Comment by Ashald** [comment]
We have similar requirements where we need to manually de-register Consul server when we destroy the ASG with them.

**Comment by saurabh-hirani** [comment]
+1 - this is a must have for different stages - even if it rudimentary script execution hooks at various points in the infra lifecycle.

**Comment by saurabh-hirani** [comment]
I am relatively new to terraform but building on @icebourg 's example and https://github.com/hashicorp/terraform/issues/2831 one could do the following

```
terraform apply -var="deregister=1" -target=null_resource.deregister_my_resource
```

with 

```
variable "deregister" {
  description = "Deregister from chef - 1 or 0"
  default = 0
}
```

and

```
resource "null_resource" "deregister_my_resource" {
  count = "${var.deregister}"
  provisioner "local-exec" {
    command = "./bin/deregister-chef.sh my_resource, var.name_lookup_key)}; true"
  }
}
```

with ./bin/deregister-chef.sh containing

```
#!/bin/bash
knife node delete -y $1
knife client delete -y $1
```

That way this can be called post a destroy.

**Comment by 5tefan** [comment]
Another use case: when removing a Kuberenetes worker, it would be nice to move pods off the node and remove it before the resource is destroyed: 
```
kubectl drain <nodename>
kubectl delete node <nodename>
```
Thanks @DocBradfordSoftware hermanjunge/kubernetes-digitalocean-terraform#21

**Comment by amclain** [comment]
>Another use case: when removing a Kuberenetes worker

I had [the same issue except with Nomad](https://gitter.im/hashicorp-nomad/Lobby?at=58177616482c168b22d6c440). When Terraform would recreate the instance(s), the containers wouldn't migrate to the live nodes until `nomad node-drain` was run, which I ended up having to do manually.

**Comment by ajbouh** [comment]
I also need it for nomad and kubernetes

**Comment by mitchellh** [comment]
I know there is a lot of discussion here but this is identical to #386 and there is an equally large amount of discussion there so I'm going to close this and centralize there.
